### PR TITLE
feat: introduce cluster limit

### PR DESCRIPTION
**Pull request by hzxa21**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Related discussion: https://github.com/risingwavelabs/risingwave/issues/16092

To ensure smoother user experience when cluster resource is insufficient, this PR introduces cluster limit in kernel. At the 1st stage, the cluster limit only contains limit on actor count per worker parallelism and there are a lot of headroom for the default limit. The behavior is:
- soft limit is configured via meta config (default to 100)
- hard limit is configured via meta config (default to 400)
- limit can be bypassed via session config

When the soft limit is hit, CREATE MV/TABLE/SINK can succeed with a notice message attached.
When the hard limit is hit, CREATE MV/TABLE/SINK cannot succeed with a notice message attached.

Example:
```
dev=> create materialized view test_cluster_limit as select * from test;
CREATE_MATERIALIZED_VIEW

dev=> create materialized view test_cluster_limit_exceed_soft as select * from test;
NOTICE:  
- Actor count per parallelism exceeds the recommended limit.
- Depending on your workload, this may overload the cluster and cause performance/stability issues. Scaling the cluster is recommended.
- Contact us via slack or https://risingwave.com/contact-us/ for further enquiry.
- You can bypass this check via SQL `SET bypass_cluster_limits TO true`.
- You can check actor count distribution via SQL `SELECT * FROM rw_worker_actor_count`.
ActorCountPerParallelism { critical limit: 8, recommended limit: 7. worker_id_to_actor_count: ["1 -> WorkerActorCount { actor_count: 32, parallelism: 4 }"] }
CREATE_MATERIALIZED_VIEW

dev=> create materialized view test_cluster_limit_exceed_hard as select * from test;
ERROR:  Failed to run the query

Caused by:
  Protocol error: 
- Actor count per parallelism exceeds the critical limit.
- Depending on your workload, this may overload the cluster and cause performance/stability issues. Please scale the cluster before proceeding!
- Contact us via slack or https://risingwave.com/contact-us/ for further enquiry.
- You can bypass this check via SQL `SET bypass_cluster_limits TO true`.
- You can check actor count distribution via SQL `SELECT * FROM rw_worker_actor_count`.
ActorCountPerParallelism { critical limit: 7, recommended limit: 6. worker_id_to_actor_count: ["1 -> WorkerActorCount { actor_count: 32, parallelism: 4 }"] }
```

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

A limit on actor count per parallelism (i.e. CPU core) is added to warn user on `CREATE MATERIALIZED VIEW/TABLE/SINK` statements to ensure smoother user experience when cluster resource is insufficient. Both a soft and hard limit are introduced:
- `meta_actor_cnt_per_worker_parallelism_soft_limit` (default to 100): when the limit is exceeded,  `CREATE MATERIALIZED VIEW/TABLE/SINK` can still succeed but a SQL notice message will be included when the statement returns.
- `meta_actor_cnt_per_worker_parallelism_hard_limit` (default to 400): when the limit is exceeded,  `CREATE MATERIALIZED VIEW/TABLE/SINK` will fail and a error message will be included when the statement returns.

Notes:
- The limit will not affect existing created database object in the cluster. In other words, there will be no disruption on MVs/TABLEs/SINKs that are already created. It will only affect future `CREATE MATERIALIZED VIEW/TABLE/SINK`.
- Scaling the cluster or reducing the streaming job parallelism is the preferred way to resolve the limits.
- The default limits are subjective to change in future releases.
- If users believe it is safe for the cluster to run with the limits exceeded, there are two ways to override the behavior
  - Bypass the check via session variable: `SET bypass_cluster_limits TO true`.
  - Increase the limit via meta developer config:
  ```toml
  [meta.developer]
  meta_actor_cnt_per_worker_parallelism_soft_limit = 100
  meta_actor_cnt_per_worker_parallelism_hard_limit = 400
  ```
  However, keep in mind the cluster can be overload with potential degradation in stability/availability/performance if the check is bypass or the limits are manually increased. Please operate at your own risk with cautions.

### CODE REVIEW

**Review by lmatz — APPROVED** [review]

**Comment by hzxa21** [comment]
Example for free-tier:
```
> create table t_free_tier_exceed_soft(a int);
NOTICE:  
- Actor count per parallelism exceeds the recommended limit.
- This may overload the cluster and cause performance/stability issues. Scaling the cluster is recommended.
- Contact us via https://risingwave.com/contact-us/ and consider upgrading your free-tier license to enhance performance/stability and user experience for production usage.
- You can check actor count distribution via SQL `SELECT * FROM rw_worker_actor_count`.
ActorCountPerParallelism { critical limit: 1, recommended limit: 0. worker_id_to_actor_count: ["1 -> WorkerActorCount { actor_count: 4, parallelism: 4 }"] }
CREATE_TABLE


> create table t_free_tier_exceed_hard(a int);
ERROR:  Failed to run the query

Caused by:
  Protocol error: 
- Actor count per parallelism exceeds the critical limit.
- This may overload the cluster and cause performance/stability issues. Please scale the cluster before proceeding.
- Contact us via https://risingwave.com/contact-us/ and consider upgrading your free-tier license to enhance performance/stability and user experience for production usage.
- You can check actor count distribution via SQL `SELECT * FROM rw_worker_actor_count`.
ActorCountPerParallelism { critical limit: 1, recommended limit: 0. worker_id_to_actor_count: ["1 -> WorkerActorCount { actor_count: 6, parallelism: 4 }"] }


```

File: proto/meta.proto [hunk 1]
```diff
@@ -789,3 +789,31 @@ message RelationIdInfos {
...
+message ActorCountPerParallelism {
+  map<uint32, WorkerActorCount> worker_id_to_actor_count = 1;
+  uint64 hard_limit = 2;
+  uint64 soft_limit = 3;
+}
+
+message ClusterLimit {
+  oneof limit {
+    ActorCountPerParallelism actor_count = 1;
+    // TODO: limit DDL using compaction pending bytes
```
Comment (fuyufjh) [inline proto/meta.proto:807]: Are you sure they are either-or relation?

File: src/meta/service/src/cluster_limit_service.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        }
+    }
+
+    async fn get_active_actor_limit(&self) -> MetaResult<Option<ClusterLimit>> {
+        let (soft_limit, hard_limit) = match LicenseManager::get().tier() {
+            Ok(Tier::Paid) => (
+                self.env.opts.actor_cnt_per_worker_parallelism_soft_limit,
+                self.env.opts.actor_cnt_per_worker_parallelism_hard_limit,
+            ),
+            Ok(Tier::Free) => (
```
Comment (fuyufjh) [inline src/meta/service/src/cluster_limit_service.rs:51]: It seems still a bit controversial to decide whether & how we should set a different limit for free users. Thus, I tend to use the same limit for both paid and free users for now until we have clearer ideas in the future.

**Review by fuyufjh — APPROVED** [review]
LGTM for the rest

File: src/meta/service/src/cluster_limit_service.rs (hunk 2 above)
Comment (hzxa21) [reply src/meta/service/src/cluster_limit_service.rs]: +1. I  have updated the PR to unify the behavior for different tiers.

File: src/common/src/session_config/mod.rs [hunk 3]
```diff
@@ -318,6 +327,18 @@ fn check_bytea_output(val: &str) -> Result<(), String> {
...
+fn check_bypass_cluster_limits(_val: &bool) -> Result<(), String> {
+    match LicenseManager::get()
+        .tier()
+        .map_err(|e| e.to_report_string())?
+    {
+        risingwave_license::Tier::Free => {
+            Err("Bypassing cluster limits is only allowed in paid tier".to_string())
+        }
+        risingwave_license::Tier::Paid => Ok(()),
+    }
```
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/common/src/session_config/mod.rs]: We have decided to unify the behavior for both tiers first.

File: src/meta/service/src/cluster_limit_service.rs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                FREE_TIER_ACTOR_CNT_SOFT_LIMIT,
+                FREE_TIER_ACTOR_CNT_HARD_LIMIT,
+            ),
```
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/meta/service/src/cluster_limit_service.rs]: ditto

File: src/frontend/src/session.rs [hunk 5]
```diff
@@ -1194,6 +1196,55 @@ impl SessionImpl {
...
+        for limit in limits {
+            match limit {
+                cluster_limit::ClusterLimit::ActorCount(l) => {
+                    if l.exceed_hard_limit() {
+                        let mut msg = "\n- Actor count per parallelism exceeds the critical limit."
+                            .to_string();
+                        msg.push_str("\n- This may overload the cluster and cause performance/stability issues. Please scale the cluster before proceeding.");
+                        if matches!(tier, risingwave_common::license::Tier::Free) {
+                            msg.push_str("\n- Contact us via https://risingwave.com/contact-us/ and consider upgrading your free-tier license to enhance performance/stability and user experience for production usage.");
+                        }
```
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/frontend/src/session.rs]: ditto

**Pull request merged**